data/org.freedesktop.portal.FileChooser.xml

```diff
@@ -273,6 +273,15 @@
             </para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term>current_folder ay</term>
+          <listitem>
+            <para>
+              Suggested folder to select the files from. The byte array is
+              expected to be null-terminated.
```

Comment:
The documentation would be a lot more useful if it stated what filesystem view will be used to resolve this path - is it a host path ? Or something visible inside the sandbox?

It would also be good to clarify 'suggested' a bit more. I am reading it to mean:
- portal implementations may ignore this
- this is not meant to limit the users freedom to provide whatever file he chooses (e.g. it is fine for him/her to navigate to another folder before making a selection)

Reply:
Also, aren't byte arrays always null-terminated?

Reply:
Though, I don't know why we want to require null termination of the byte arrays.  The null at the end of a string is not considered to be part of the string itself, so similar logic dictates that byte arrays should not semantically include the null byte.  (Whether or not the marshalling format will place a null at the end of the byte array is a different matter, but I don't think D-Bus users should not rely on that.)

Reply:
> The documentation would be a lot more useful if it stated what filesystem view will be used to resolve this path - is it a host path ? Or something visible inside the sandbox?

I agree that to make this interface useful, the documentation needs to answer this question.

> It would also be good to clarify 'suggested' a bit more

I agree.


> Also, aren't byte arrays always null-terminated?

D-Bus doesn't require byte arrays to be zero-terminated. A byte array can be any arbitrary blob of bytes, similar to any other integer array but with 1-byte elements; it's like a `struct { unsigned char *location, size_t length }` in C.

As a higher-level, application-layer thing, an interface can define specific semantics for a byte array, like "it contains the bytes of a valid JPEG". In this particular case, the interface is using the convention for representing "bytestrings" (conceptually a string in an arbitrary ASCII superset, similar to `(type filename)` in GObject-Introspection), which is to encode them in the D-Bus message as a byte array with length `strlen(str) + 1`, containing the `strlen(str)` non-zero bytes of the string, followed by exactly one byte with value zero.

In the D-Bus Specification, the LinuxSecurityLabel is currently the only standardized bytestring, and is described as "the non-zero bytes of [whatever] in an unspecified ASCII-compatible encoding, followed by a single zero byte". I think that's a good way to phrase it.

"Null-terminated" is a somewhat misleading term for this, because this is byte value 0 (ASCII mnemonic `NUL`, not a `NULL` pointer.

It's correct to use the bytestring convention for Unix filenames, environment variable names/values, and any similar string-like thing that (unfortunately) cannot be guaranteed to be UTF-8 (and might instead be a legacy national character set like Latin-1).

> Whether or not the marshalling format will place a null at the end of the byte array is a different matter

The underlying marshalling format does not guarantee that there will be a zero after the end of a byte array, and also does not guarantee that there will not be a zero in the middle of the byte array. The higher-level (application-level) bytestring convention is that there is exactly one zero in the length-counted content of the array, as the last item.